pkg/errors.go

```diff
@@ -0,0 +1,156 @@
+//
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitops
+
+import (
+	"fmt"
+
+	"github.com/redhat-developer/gitops-generator/pkg/util"
+)
+
+type GitCmd string
+
+const (
+	cloneRepo      GitCmd = "clone git"
+	checkGitDiff   GitCmd = "check git diff"
+	commitFiles    GitCmd = "commit files"
+	pushRemote     GitCmd = "push remote"
+	initializeGit  GitCmd = "initialize git"
+	addComponents  GitCmd = "add components"
+	getCommitID    GitCmd = "retrieve commit id"
+	switchBranch   GitCmd = "switch to"
+	checkoutBranch GitCmd = "checkout"
+	genOverlays    GitCmd = "overlays dir"
+)
+
+// GitCmdError is used to construct custom errors for a number of git commands that follow similar message patterns
+// Used by the following command types:  cloneRepo, checkGitDiff, commitFiles, pushRemote, initializeGit, addComponents, getCommitID
+
+type GitCmdError struct {
+	path      string
+	cmdResult string
+	err       error
+	cmdType   GitCmd
+}
+
+func (e *GitCmdError) Error() string {
+
+	// Check to see what gitops actions have take place in order to append the correct preposition
+	cmdMsg := e.cmdType
+	if e.cmdType == checkGitDiff {
+		cmdMsg = cmdMsg + " in"
+	} else if e.cmdType == commitFiles || e.cmdType == pushRemote {
+		cmdMsg = cmdMsg + " to"
+	} else if e.cmdType == getCommitID {
+		cmdMsg = cmdMsg + " for"
+	}
+
+	return util.SanitizeErrorMessage(fmt.Errorf("failed to %s repository %q %q: %s", cmdMsg, e.path, e.cmdResult, e.err)).Error()
+}
+
+// GitBranchError is used to construct custom errors related to git branch failures
+// Used by the following command types: switchBranch, checkoutBranch
+type GitBranchError struct {
+	branch    string
+	repoPath  string
+	cmdResult string
+	err       error
+	cmdType   GitCmd
+}
+
+func (e *GitBranchError) Error() string {
+	return util.SanitizeErrorMessage(fmt.Errorf("failed to %s branch %q in repository %q %q: %s", e.cmdType, e.branch, e.repoPath, string(e.cmdResult), e.err)).Error()
+}
+
+type GitGenResourcesAndOverlaysError struct {
+	path          string
+	componentName string
+	err           error
+	cmdType       GitCmd
+}
+
+func (e *GitGenResourcesAndOverlaysError) Error() string {
+	failedToGenerate := "failed to generate the gitops resources in "
+	if e.cmdType == genOverlays {
```

Comment:
Could genOverlays be "generate overlays directory"
and then  we don't need this if block, and modify the message on line 90 to be:
"failed to %q for component %q: %s" 

That way the message on line 90 is similar to the other messages like on line 75, and 102, etc.?  They all start with "failed to...."

Reply:
yeah, I did start off with a similar solution but the substitution leaves an extra space if it's not an "overlays dir".  This breaks unit testing